mixing/mixpool/mixpool.go

```diff
@@ -0,0 +1,1523 @@
+
+	for _, tx := range txs {
+		txHash := tx.TxHash()
+		ses, ok := p.sessionsByTxHash[txHash]
+		if ok {
+			p.removeSession(ses.sid, &txHash, true)
+			continue
+		}
+
+		for _, in := range tx.TxIn {
+			prHash := p.outPoints[in.PreviousOutPoint]
+			pr, ok := p.prs[prHash]
+			if ok {
+				p.removePR(pr, "double spent")
+			}
+		}
+	}
+}
+
+// ReceiveKEsByPairing returns the most recently received run-0 KE messages by
+// a peer that reference PRs of a particular pairing and epoch.
+func (p *Pool) ReceiveKEsByPairing(pairing []byte, epoch uint64) []*wire.MsgMixKeyExchange {
+	p.mtx.RLock()
+	defer p.mtx.RUnlock()
+
+	var kes []*wire.MsgMixKeyExchange
+	for id, ke := range p.latestKE {
+		prHash := p.messagesByIdentity[id][0]
+		pr := p.prs[prHash]
+		prPairing, err := pr.Pairing()
+		if err != nil {
+			continue
+		}
+		if bytes.Equal(pairing, prPairing) && ke.Epoch == epoch {
+			kes = append(kes, ke)
+		}
+	}
+	return kes
+}
+
+// RemoveUnresponsiveDuringEpoch removes pair requests of unresponsive peers
+// that did not provide any key exchange messages during the epoch in which a
+// mix occurred.
+func (p *Pool) RemoveUnresponsiveDuringEpoch(prs []*wire.MsgMixPairReq, epoch uint64) {
+	p.mtx.Lock()
+	defer p.mtx.Unlock()
+
+	var unresponsive []*wire.MsgMixPairReq
+
+PRLoop:
+	for _, pr := range prs {
+		for _, msgHash := range p.messagesByIdentity[pr.Identity] {
+			msg, ok := p.pool[msgHash].msg.(*wire.MsgMixKeyExchange)
+			if !ok {
+				continue
+			}
+			if msg.Epoch == epoch {
+				continue PRLoop
+			}
+		}
+
+		unresponsive = append(unresponsive, pr)
+	}
+
+	for _, pr := range unresponsive {
+		p.removePR(pr, "unresponsive")
+	}
+}
+
+// Received is a parameter for Pool.Receive describing the session and run to
+// receive messages for, and slices for returning results.  Only non-nil slices
+// will be appended to.  Received messages are unsorted.
+type Received struct {
+	Sid [32]byte
+	Run uint32
+	KEs []*wire.MsgMixKeyExchange
+	CTs []*wire.MsgMixCiphertexts
+	SRs []*wire.MsgMixSlotReserve
+	DCs []*wire.MsgMixDCNet
+	CMs []*wire.MsgMixConfirm
+	FPs []*wire.MsgMixFactoredPoly
+	RSs []*wire.MsgMixSecrets
+}
+
+// Receive returns messages matching a session, run, and message type, waiting
+// until all described messages have been received, or earlier with the
+// messages received so far if the context is cancelled before this point.
+//
+// Receive only returns results for the session ID and run increment in the r
+// parameter.  If no such session or run has any messages currently accepted
+// in the mixpool, the method immediately errors.
+//
+// If any secrets messages are received for the described session and run, and
+// r.RSs is nil, Receive immediately returns ErrSecretsRevealed.  An
+// additional call to Receive with a non-nil RSs can be used to receive all of
+// the secrets after each peer publishes their own revealed secrets.
+func (p *Pool) Receive(ctx context.Context, expectedMessages int, r *Received) error {
```

Comment:
It is only used to receive messages of one type at a time, so the expected number of messages is accurate.

We could error for unexpected uses of this, where multiple non-nil slices are passed in.

Reply:
That would work.  That expected usage wasn't clear from the comment because it says "waiting until all described messages have been received" along with "Only non-nil slices will be appended to." which I interpreted to mean whatever mix of non-nil slices you provide should be filled and then that is what the code actually does too.

Reply:
maybe generics would improve this api, like Receive[*wire.MsgMixKE] or something, but i have not tried to figure out how to make that work well (or at all).

Reply:
I am not sure if changing the expectedMessages param to anything but int will be beneficial.  Total message counts are already highly restricted by how many peers can participate in a single mix session, which wire caps to 512, and all of the current uses of this api in mixclient pass in ints in obvious ways (always a len() directly, or in a single case for receiving FP messages, there is a loop with an increasing expectedMessages variable, always set to `len(received.FPs) + 1` before the next iteration).